### DISCUSSION TITLE
Rationale for having Long* and Double* variants of all metrics types

### DISCUSSION THREAD

**Discussion by shakuzen** [open-telemetry/opentelemetry-java#2535] [Q&A]
I'm trying to understand the necessity behind there being Long* and Double* variants of every [Metrics Instruments API](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/metrics/api.md#metric-instruments). I notice this isn't part of the general API specification and is specific to the Java implementation. I was not able to find any documented rationale explaining this decision in this repository. If I've missed it, could someone point me at it? If it doesn't exist yet, I think it would be good to document that somewhere as surely the question will come up. If nothing else, every user is faced with making the decision of which instrument variant to use every time.

/cc @jonatan-ivanov

**Comment by anuraaga** [comment]
Hey @shakuzen - I'm not sure of any rationale explanation, but the spec for having long and double is here I think

https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/metrics/api.md#metric-event-format

`value (signed integer or floating point number)`

**Reply by shakuzen** [reply]
Thanks for the quick response.

Perhaps my reading of that part of the spec is different from other peoples' but that doesn't seem to necessitate there being two variants of each type. If a Java `double` `value` is accepted as input, that allows a signed integer (`long`) or a floating point number (`double`) to be passed, fulfilling the requirements of the spec. Unless the spec means the internal representation of the measurement has to cover both, which seems odd in an API specification, and then I guess I need to ask for clarification/rationale to the API specification instead of the Java implementation.

**Reply by anuraaga** [reply]
That's a good point. There's still much work left on the spec from what I see so hopefully things get cleaned up.

For anyone joining tomorrow's workshop (@jkwatson) my vote goes to double-only. When first starting to use prometheus, it was jarring to only have double. But after using it, I realized how it was never really limiting and just generally a lot simpler :)

**Reply by jkwatson** [reply]
Ping @bogdandrutu on this, as I won't be able to attend much of the workshop, due to conflicts.
